src/main/java/org/opentripplanner/graph_builder/module/osm/naming/SidewalkNamer.java

```diff
@@ -0,0 +1,304 @@
+import org.locationtech.jts.geom.LineString;
+import org.locationtech.jts.geom.MultiLineString;
+import org.locationtech.jts.geom.Point;
+import org.locationtech.jts.operation.buffer.BufferParameters;
+import org.locationtech.jts.operation.distance.DistanceOp;
+import org.opentripplanner.framework.geometry.GeometryUtils;
+import org.opentripplanner.framework.geometry.HashGridSpatialIndex;
+import org.opentripplanner.framework.geometry.SphericalDistanceLibrary;
+import org.opentripplanner.framework.i18n.I18NString;
+import org.opentripplanner.framework.lang.DoubleUtils;
+import org.opentripplanner.framework.logging.ProgressTracker;
+import org.opentripplanner.graph_builder.module.osm.StreetEdgePair;
+import org.opentripplanner.graph_builder.services.osm.EdgeNamer;
+import org.opentripplanner.openstreetmap.model.OSMWithTags;
+import org.opentripplanner.street.model.edge.StreetEdge;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A namer that assigns names of nearby streets to sidewalks if they meet certain
+ * geometric similarity criteria.
+ * <p>
+ * The algorithm works as follows:
+ *  - for each sidewalk we look up (named) street edges nearby
+ *  - group those edges into groups where each edge has the same name
+ *  - draw a flat-capped buffer around the sidewalk, like this: https://tinyurl.com/4fpe882h
+ *  - check how much of a named edge group is inside the buffer
+ *  - remove those groups which are below MIN_PERCENT_IN_BUFFER
+ *  - take the group that has the highest percentage (as a proportion of the sidewalk length) inside
+ *    the buffer and apply its name to the sidewalk.
+ * <p>
+ * This works very well for OSM data where the sidewalk runs a parallel to the street and at each
+ * intersection the sidewalk is also split. It doesn't work well for sidewalks that go around
+ * the corner, like https://www.openstreetmap.org/way/1059101564. These cases are, however, detected
+ * by the above algorithm and the sidewalk name remains the same.
+ */
+public class SidewalkNamer implements EdgeNamer {
+
+  private static final Logger LOG = LoggerFactory.getLogger(SidewalkNamer.class);
+  private static final int MAX_DISTANCE_TO_SIDEWALK = 50;
+  private static final double MIN_PERCENT_IN_BUFFER = .85;
+  private static final int BUFFER_METERS = 25;
+
+  private HashGridSpatialIndex<EdgeOnLevel> streetEdges = new HashGridSpatialIndex<>();
+  private Collection<EdgeOnLevel> unnamedSidewalks = new ArrayList<>();
+  private PreciseBuffer preciseBuffer;
+
+  @Override
+  public I18NString name(OSMWithTags way) {
+    return way.getAssumedName();
+  }
+
+  @Override
+  public void recordEdges(OSMWithTags way, StreetEdgePair pair) {
+    if (way.isSidewalk() && way.needsFallbackName() && !way.isExplicitlyUnnamed()) {
+      pair
+        .asIterable()
+        .forEach(edge -> unnamedSidewalks.add(new EdgeOnLevel(edge, way.getLevels())));
+    }
+    if (way.isNamed() && !way.isLink()) {
```

Comment:
If the edges for a sidewalk way have been added to unnamedSidewalks in the conditional above, does this second conditional allow those edges to also be added to the spatial index? That spatial index is for finding roads close to sidewalks, so it shouldn't contain sidewalks right?

On further inspection I think I see what's going on. The needsFallbackName method is defined as `! isNamed()` so both conditionals cannot evaluate to true for a single way. But for the reader/maintainer it is not at all obvious that something that "needs a fallback name" simply does not have a name. Could this be structured as an `if ... else if ...` to make it obvious and foolproof that these are mutually exclusive conditional blocks, with comments labeling them as the unnamed sidewalk case and the named road case? I'm also wondering if the method should just be called `way.isNotNamed()` to make it more obvious that it's just syntactic sugar for `!way.isNamed()`.

Reply:
> If the edges for a sidewalk way have been added to unnamedSidewalks in the conditional above, does this second conditional allow those edges to also be added to the spatial index? That spatial index is for finding roads close to sidewalks, so it shouldn't contain sidewalks right?
> 
> On further inspection I think I see what's going on. The needsFallbackName method is defined as `! isNamed()` so both conditionals cannot evaluate to true for a single way. But for the reader/maintainer it is not at all obvious that something that "needs a fallback name" simply does not have a name. Could this be structured as an `if ... else if ...` to make it obvious and foolproof that these are mutually exclusive conditional blocks, with comments labeling them as the unnamed sidewalk case and the named road case? 

I've done that in e9a0722d9b6c4a6e28a48ca85a799fb4ece45e62

> I'm also wondering if the method should just be called `way.isNotNamed()` to make it more obvious that it's just syntactic sugar for `!way.isNamed()`.

I'm a little bit reluctant about that change because there are ways that simply have no name in OSM but there are also sidewalks which really, really don't have a name and are tagged as `noname=yes`. Those ones are checked with `isExplicitlyUnnamed`. Maybe we have a quick check in the dev meeting.

Reply:
Thanks for the else-if, I do find it easier to comprehend this way. 

I now see what you mean about `isNotNamed` considering there is a possibility of being explicitly unnamed. However, I'm not sure when someone (like me) sees `needsFallbackName` they'll understand this nuance, and might get stuck on the new concept of a "fallback name" which appears to be somehow equivalent or related to a "non-bogus name". It would probably benefit from an explanation. There are other options like like `hasName()` and `hasNoName()` or just `! hasName()`. But I don't want to get stuck on this point - if you like the existing naming we can just keep it.

Reply:
On reflection the concept of fallback name which I introduced here probably does more harm than good. Whether something really needs a fallback name is quite subjective and cannot be derived from the source data.

For that reason I went with the follow method names:

- hasName
- hasNoName
- isExplicitlyUnnamed

This strictly states what can be seen from the source data and not some form of other judgement. I've also clarified the Javadoc.

You touched on another point which lead me into this in the first place: I had to google what "bogus name" meant, since it's not a commen word in British English. I would like to replace it with another term. Maybe we can come up with better name in a dev meeting.